### ISSUE TITLE
regression "Cannot read property 'registerOnChange' of null at cleanUpControl" / "Error during cleanup of component"

### ISSUE THREAD

**Issue by satanTime** [angular/angular#40521]
# 🐞 bug report

### Affected Package
@angular/core/testing **v11.1.0**

### Is this a regression?

Yes, the previous version in which this bug was not present was: v11.0.9

### Description

tests for reactive forms without value accessor fail in clean up.

## 🔬 Minimal Reproduction

https://codesandbox.io/s/twilight-thunder-tbl3d?file=/src/test.spec.ts (open console)
change angular to 11.0.9 to see that error isn't present: https://take.ms/wuPcy

## 🔥 Exception or Error
<pre><code>
Error during cleanup of component
</code></pre>

which is caused by

<pre><code>
Cannot read property 'registerOnChange' of null at cleanUpControl
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/

Angular CLI: 11.1.0
Node: 12.20.1
OS: darwin x64

Angular: 11.1.0
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1101.0
@angular-devkit/build-angular   0.1101.0
@angular-devkit/core            11.1.0
@angular-devkit/schematics      11.1.0
@schematics/angular             11.1.0
@schematics/update              0.1101.0
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

**Comment by JoostK** [comment]
Here's the full stacktrace:

```
TypeError: Cannot read property 'registerOnChange' of null
    at cleanUpControl (https://tbl3d.csb.app/node_modules/@angular/forms/fesm2015/forms.js:1664:21)
    at FormGroupDirective.removeControl (https://tbl3d.csb.app/node_modules/@angular/forms/fesm2015/forms.js:3189:5)
    at FormControlName.ngOnDestroy (https://tbl3d.csb.app/node_modules/@angular/forms/fesm2015/forms.js:3494:26)
    at callProviderLifecycles (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:16098:14)
    at callElementProvidersLifecycles (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:16069:7)
    at callLifecycleHooksChildrenFirst (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:16059:21)
    at destroyView (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:20501:3)
    at callViewAction (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:20609:7)
    at execComponentViewsAction (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:20546:7)
    at destroyView (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:20500:3)
    at callWithDebugContext (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:21134:23)
    at Object.debugDestroyView [as destroyView] (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:20899:10)
    at ViewRef_.destroy (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:15584:14)
    at ComponentRef_.destroy (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/core.js:15433:19)
    at ComponentFixture.destroy (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/testing.js:308:25)
    at eval (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/testing.js:1602:17)
    at Array.forEach (<anonymous>)
    at TestBedViewEngine.resetTestingModule (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/testing.js:1600:26)
    at Function.resetTestingModule (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/testing.js:1509:29)
    at UserContext.eval (https://tbl3d.csb.app/node_modules/@angular/core/fesm2015/testing.js:1913:13)
    at ZoneDelegate.invoke (https://tbl3d.csb.app/node_modules/zone.js/fesm2015/zone.js:368:26)
    at ProxyZoneSpec.onInvoke (eval at z (https://codesandbox.io/static/js/sandbox.d6312f59f.js:1:98947), <anonymous>:300:43)
    at ZoneDelegate.invoke (https://tbl3d.csb.app/node_modules/zone.js/fesm2015/zone.js:367:52)
    at Zone.run (https://tbl3d.csb.app/node_modules/zone.js/fesm2015/zone.js:130:43)
    at runInTestZone (eval at z (https://codesandbox.io/static/js/sandbox.d6312f59f.js:1:98947), <anonymous>:580:38)
    at UserContext.eval (eval at z (https://codesandbox.io/static/js/sandbox.d6312f59f.js:1:98947), <anonymous>:595:24)
    at ZoneQueueRunner.attempt (https://tbl3d.csb.app/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:7063:44)
    at ZoneQueueRunner.QueueRunner.run (https://tbl3d.csb.app/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:7104:25)
    at runNext (https://tbl3d.csb.app/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:7023:18)
    at next (https://tbl3d.csb.app/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:7030:11)
    at eval (https://tbl3d.csb.app/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:6924:9)
    at ZoneDelegate.invokeTask (https://tbl3d.csb.app/node_modules/zone.js/fesm2015/zone.js:402:31)
    at Zone.runTask (https://tbl3d.csb.app/node_modules/zone.js/fesm2015/zone.js:174:47)
    at drainMicroTaskQueue (https://tbl3d.csb.app/node_modules/zone.js/fesm2015/zone.js:578:35)
```

Notice how this uses ViewEngine, not Ivy, but that may not be relevant.

**Comment by AndrewKushnir** [comment]
Hi @satanTime, thanks for reporting the issue.

PR #39235 (changes were released in v11.1.0) introduced additional cleanup logic for form controls and directives. This cleanup logic relies on the fact that `FormControlName` and `FormControl` directives are fully initialized, thus having the `valueAccessor` field defined as an instance of corresponding CVA (Control Value Accessor). There is a logic in [selectValueAccessor function](https://github.com/angular/angular/blob/master/packages/forms/src/directives/shared.ts#L335) that makes sure that this is the case and throws an error otherwise. However in your test there is a try-catch (in a form of `expect(...).toThrow(...)`) which ignores that error, thus directive instance ended up being not fully initialized.

This is unlikely to happen in real scenarios (i.e. in an app on a page), but some specific use-cases in tests might cause the error to happen. So I've created a PR #40526 to make the cleanup logic more resilient to this type of scenarios.

Thank you.

**Comment by satanTime** [comment]
Hi.

Totally agree. We have a lib for testing which contains a couple of tests to ensure that it fails in the same way as angular does, but not sure that someone would do something similar in a real project.

Thank you and happy coding!

**Comment by akram1905** [comment]
Hi,
It actually broke our app on production! 
_[Partially it was our fault we didn't lock Angular's version on package.json]_

We use a directive to add dynamic Form controls, then in `ngOnDestroy` we call `removeControl` which caused the error.

**Comment by satanTime** [comment]
Ah, here we go, my condolences.

**Comment by akram1905** [comment]
Thank you for your kindness and sympathy.
I’m deeply moved by your comforting words.
I greatly appreciate your precious gifts that you are thinking to send me since as you know words won't be enough to condolence me.

**Comment by AndrewKushnir** [comment]
Hi @akram1905, thanks for providing additional information.

The mentioned PR #40526 is in the merge queue and the fix will be available in v11.1.1 later this week.

> We use a directive to add dynamic Form controls, then in ngOnDestroy we call removeControl which caused the error.

The code that was there before in `cleanUpControl` function always expected a CVA instance to be present on a directive and there is also a code in `selectValueAccessor` that throws if it's not the case. I'm wondering if you could share a minimal repro of your use-case (via [Stackblitz](https://stackblitz.com/))? I'd like to make sure we have test coverage for it and/or add more tests if needed to prevent regressions in the future. Note: this issue will be closed automatically once #40526 is merged, but you could still leave a comment with additional details in this thread.

Thank you.
